src/orion/client/runner.py

```diff
@@ -0,0 +1,344 @@
+    CompletedExperiment,
+    InvalidResult,
+    ReservationRaceCondition,
+    ReservationTimeout,
+    WaitingForTrials,
+)
+from orion.core.utils.flatten import flatten, unflatten
+from orion.core.worker.consumer import ExecutionError
+from orion.core.worker.trial import AlreadyReleased
+from orion.executor.base import AsyncException, AsyncResult
+
+
+class LazyWorkers(Exception):
+    """Raised when all the workers have been idle for a given amount of time"""
+
+    pass
+
+
+def _optimize(trial, fct, trial_arg, **kwargs):
+    """Execute a trial on a worker"""
+    kwargs.update(flatten(trial.params))
+
+    if trial_arg:
+        kwargs[trial_arg] = trial
+
+    return fct(**unflatten(kwargs))
+
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class _Stat:
+    sample: int = 0
+    scatter: int = 0
+    gather: int = 0
+
+    @contextmanager
+    def timer(self, name):
+        """Measure elapsed  time of a given block"""
+        start = time.time()
+        yield
+        total = time.time() - start
+
+        value = getattr(self, name)
+        setattr(self, name, value + total)
+
+    def report(self):
+        """Show the elapsed time of different blocks"""
+        lines = [
+            f"Sample  {self.sample:7.4f}",
+            f"Scatter {self.scatter:7.4f}",
+            f"Gather  {self.gather:7.4f}",
+        ]
+        return "\n".join(lines)
+
+
+class Runner:
+    """Run the optimization process given the current executor"""
+
+    def __init__(
+        self,
+        client,
+        fct,
+        pool_size,
+        idle_timeout,
+        max_trials_per_worker,
+        max_broken,
+        trial_arg,
+        on_error,
+        interrupt_signal_code=None,
+        gather_timeout=0.01,
+        **kwargs,
+    ):
+        self.client = client
+        self.fct = fct
+        self.pool_size = pool_size
+        self.max_trials_per_worker = max_trials_per_worker
+        self.max_broken = max_broken
+        self.trial_arg = trial_arg
+        self.on_error = on_error
+        self.kwargs = kwargs
+
+        self.gather_timeout = gather_timeout
+        self.idle_timeout = idle_timeout
+
+        self.worker_broken_trials = 0
+        self.trials = 0
+        self.futures = []
+        self.pending_trials = dict()
+        # self.free_worker = pool_size
+        self.stat = _Stat()
+
+        if interrupt_signal_code is None:
+            interrupt_signal_code = orion.core.config.worker.interrupt_signal_code
+
+        self.interrupt_signal_code = interrupt_signal_code
+
+    @property
+    def free_worker(self):
+        """Returns the number of free worker"""
+        return min(
+            max(self.client.executor.n_workers - len(self.pending_trials), 0),
+            # This comes from `workon(n_workers)` it limits the number of worker
+            # that can be used by the executor, even if the executor is created with
+            # 5 workers if pool_size is 2 then only 2 workers will be used
```

Comment:
There is a confusion here. The pool-size does not control the number of workers that can be used nor the maximum number of trials that can be pending in total. It controls the number of trials that the algorithm may sample per batch. When we call `suggest(pool_size)`, the algorithm will return `pool_size` samples. We can still call it more times if we require more trials to feed every workers. The pool-size is a way to trade-off between efficiency of algorithm sampling versus quality of samples.  Large pool-size is more efficient in wall clock time, but it may return more trials than necessary, therefore some trials were sampled with less information than would have been available when the additional were really needed. 

Maybe `batch_size` would be more clear than `pool_size`.